File: crates/planner/src/catalog.rs
//! Table catalog for the V1 SQL subset (spec 16: metrics, spans, logs,
//! deployments, incidents).

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ColumnType {
    Text,
    Int,
    Float,
    Time,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Column {
    pub name: &'static str,
    pub column_type: ColumnType,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Table {
    pub name: &'static str,
    pub columns: Vec<Column>,
}

fn col(name: &'static str, column_type: ColumnType) -> Column {
    Column { name, column_type }
}

pub fn tables() -> Vec<Table> {
    use ColumnType::*;
    vec![
        Table {
            name: "metrics",
            columns: vec![
                col("service", Text),
                col("name", Text),
                col("value", Float),
                col("event_time", Time),
            ],
        },
        Table {
            name: "spans",
            columns: vec![
                col("service", Text),
                col("operation", Text),
                col("duration_ns", Int),
                col("status", Text),
                col("trace_id", Text),
                col("event_time", Time),
            ],
        },
        Table {
            name: "logs",
            columns: vec![
                col("service", Text),
                col("severity", Text),
                col("body", Text),
                col("event_time", Time),
            ],
        },
        Table {
            name: "deployments",
            columns: vec![
                col("service", Text),
                col("change_id", Text),
                col("change_type", Text),
                col("version_after", Text),
                col("event_time", Time),
            ],
        },
        Table {
            name: "incidents",
            columns: vec![
                col("incident_id", Text),
                col("start_time", Time),
                col("end_time", Time),
            ],
        },
    ]
}

pub fn table(name: &str) -> Option<Table> {
    tables().into_iter().find(|t| t.name == name)
}

File: web/src/content/catalog.ts
// The five tables the SQL subset knows. Mirrors crates/planner/src/catalog.rs;
// there is no catalog endpoint, so this copy is the reference the workbench
// shows. Keep the two in sync when the planner grows a column.

export type CatalogColumn = { name: string; type: "text" | "float" | "int" | "time" };
export type CatalogTable = { name: string; columns: CatalogColumn[]; note: string };

export const CATALOG: CatalogTable[] = [
  {
    name: "metrics",
    note: "One row per metric sample. `name` is the series (latency, mem, cpu, error_rate).",
    columns: [
      { name: "service", type: "text" },
      { name: "name", type: "text" },
      { name: "value", type: "float" },
      { name: "event_time", type: "time" },
    ],
  },
  {
    name: "spans",
    note: "One row per span. `status` is ok or error; `trace_id` groups a request.",
    columns: [
      { name: "service", type: "text" },
      { name: "operation", type: "text" },
      { name: "duration_ns", type: "int" },
      { name: "status", type: "text" },
      { name: "trace_id", type: "text" },
      { name: "event_time", type: "time" },
    ],
  },
  {
    name: "logs",
    note: "One row per log line. `severity` is the level the source emitted.",
    columns: [
      { name: "service", type: "text" },
      { name: "severity", type: "text" },
      { name: "body", type: "text" },
      { name: "event_time", type: "time" },
    ],
  },
  {
    name: "deployments",
    note: "Change events. The right side of the temporal interval join.",
    columns: [
      { name: "service", type: "text" },
      { name: "change_id", type: "text" },
      { name: "change_type", type: "text" },
      { name: "version_after", type: "text" },
      { name: "event_time", type: "time" },
    ],
  },
  {
    name: "incidents",
    note: "The loaded incident's bounds.",
    columns: [
      { name: "incident_id", type: "text" },
      { name: "start_time", type: "time" },
      { name: "end_time", type: "time" },
    ],
  },
];

export const CATALOG_CAVEAT =
  "Time columns are event time. The planner accepts SELECT, WHERE, TUMBLE windows, P50/P95/P99, AVG, COUNT, an equi-join against deployments, ORDER BY and LIMIT; it rejects DISTINCT, subqueries and non-equi joins.";
